### PR TITLE
Syncing ARP table

### PR DESCRIPTION
**Pull request by leodotcloud**
https://github.com/rancher/rancher/issues/8206

### CODE REVIEW

File: arpsync/watcher.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+			logrus.Errorf("arpsync: while syncing, got error: %v", err)
+		}
+	}
+}
+
+func (atw *ARPTableWatcher) doSync() error {
+	logrus.Debugf("arpsync: checking the ARP table")
+	networks, err := atw.mc.GetNetworks()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching networks from metadata")
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: you aren't logging the error.  either log the error or return a wraped error with this message

File: arpsync/watcher.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	logrus.Debugf("arpsync: entries=%+v", entries)
+
+	containers, err := atw.mc.GetContainers()
+	if err != nil {
+		logrus.Errorf("arpsync: error fetching containers from metadata")
+		return err
+	}
+	containersMap, err := buildContainersMap(containers, localNetwork)
+	//logrus.Debugf("arpsync: containersMap: %v", containersMap)
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: remove commented debug statements

File: arpsync/watcher.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+					newEntry.HardwareAddr = newHardwareAddr
+					newEntry.Type = netlink.NUD_REACHABLE
+					if err := netlink.NeighSet(&newEntry); err != nil {
+						logrus.Errorf("arpsync: error changing ARP entry: %v", err)
+					}
+				}
+			}
+		} else {
+			logrus.Debugf("arpsync: container not found for ARP entry: %+v", aEntry)
+			//if err := netlink.NeighDel(&aEntry); err != nil {
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: remove

File: arpsync/watcher.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			continue
+		}
+		containersMap[aContainer.PrimaryIp] = &containers[index]
+	}
+
+	return containersMap, nil
+}
+
+func (atw *ARPTableWatcher) syncLoop() {
+
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: remove whitespace

File: arpsync/watcher.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	go atw.syncLoop()
+
+	return nil
+}
+
+func buildContainersMap(containers []metadata.Container, network metadata.Network) (map[string]*metadata.Container, error) {
+	containersMap := make(map[string]*metadata.Container)
+
+	for index, aContainer := range containers {
+		if !(aContainer.PrimaryIp != "" && aContainer.NetworkUUID == network.UUID) {
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: you should drop entries where the mac address is empty too

File: arpsync/watcher.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+		logrus.Errorf("arpsync: error fetching containers from metadata")
+		return err
+	}
+	containersMap, err := buildContainersMap(containers, localNetwork)
+	//logrus.Debugf("arpsync: containersMap: %v", containersMap)
+
+	for _, aEntry := range entries {
+		//logrus.Debugf("arpsync: aEntry: %+v", aEntry)
+		if container, found := containersMap[aEntry.IP.String()]; found {
+			if container.HostUUID == host.UUID {
```
Comment (ibuildthecloud) [inline arpsync/watcher.go]: it appears this if/else is just copy and paste but the desired mac address is different.  Can you simplify the code.

File: arpsync/watcher.go (hunk 1 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: This is being logged from the caller: https://github.com/rancher/plugin-manager/pull/60/files/544292e0c05930b8d4ceea0fc88af2ba15cfea73#diff-17ce03a91e4d814b1c8a5f001a723155R69. Isn't that sufficient?

File: arpsync/watcher.go (hunk 2 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: ok

File: arpsync/watcher.go (hunk 3 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: ok

File: arpsync/watcher.go (hunk 4 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: done.

File: arpsync/watcher.go (hunk 5 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: adding the check.

File: arpsync/watcher.go (hunk 6 above)
Comment (leodotcloud) [reply arpsync/watcher.go]: refactored.

File: arpsync/watcher.go (hunk 1 above)
Comment (ibuildthecloud) [reply arpsync/watcher.go]: Not really

**Comment by deniseschannon** [comment]
https://github.com/rancher/rancher/issues/8412

**Pull request merged**